pkg/istio/virtualservice.go

```diff
@@ -0,0 +1,145 @@
+)
+
+const (
+	// VirtualServiceKind is the k8 Kind for a istio VirtualService
+	VirtualServiceKind = "VirtualService"
+)
+
+var _ model.Mutator = VirtualServiceMutator
+var _ model.Revertor = VirtualServiceRevertor
+
+func VirtualServiceMutator(ctx model.SessionContext, ref *model.Ref) error {
+	if len(ref.GetResourceStatus(VirtualServiceKind)) > 0 {
+		return nil
+	}
+
+	targetName := strings.Split(ref.Name, "-")[0]
+
+	vs, err := getVirtualServiceMapped(ctx.Namespace, targetName)
+	//vs, err := getVirtualService(ctx, ctx.Namespace, targetName)
+	if err != nil {
+		ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: targetName, Action: model.ActionFailed})
+		return err
+	}
+
+	ctx.Log.Info("Found VirtualService", "name", targetName)
+	mutatedVs, err := mutateVirtualService(*vs)
+	if err != nil {
+		ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: targetName, Action: model.ActionFailed})
+		return err
+	}
+
+	//err = setVirtualService(ctx.Namespace, &mutatedVs)
+	err = ctx.Client.Update(ctx, &mutatedVs)
+	if err != nil {
+		ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: targetName, Action: model.ActionFailed})
+		return err
+	}
+
+	ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: targetName, Action: model.ActionModified})
+	return nil
+}
+
+func VirtualServiceRevertor(ctx model.SessionContext, ref *model.Ref) error {
+	resources := ref.GetResourceStatus(VirtualServiceKind)
+
+	for _, resource := range resources {
+		vs, err := getVirtualServiceMapped(ctx.Namespace, resource.Name)
+		//vs, err := getVirtualService2(ctx, ctx.Namespace, resource.Name)
+		if err != nil {
+			if errors.IsNotFound(err) { // Not found, nothing to clean
+				break
+			}
+			ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: resource.Name, Action: model.ActionFailed})
+			break
+		}
+
+		ctx.Log.Info("Found VirtualService", "name", resource.Name)
+		mutatedVs, err := revertVirtualService(*vs)
+		if err != nil {
+			ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: resource.Name, Action: model.ActionFailed})
+			break
+		}
+		//err = setVirtualService(ctx.Namespace, &mutatedVs)
+		err = ctx.Client.Update(ctx, &mutatedVs)
+		if err != nil {
+			ref.AddResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: resource.Name, Action: model.ActionFailed})
+			break
+		}
+		// ok, removed
+		ref.RemoveResourceStatus(model.ResourceStatus{Kind: VirtualServiceKind, Name: resource.Name})
+	}
+
+	return nil
+}
+
+func mutateVirtualService(vs istionetwork.VirtualService) (istionetwork.VirtualService, error) {
+
+	source := vs.Spec.Http[0]
+
+	sourceRoutes := []*v1alpha3.HTTPRouteDestination{}
+	for _, r := range source.Route {
+		sourceRoute := v1alpha3.HTTPRouteDestination{
+			Destination: &v1alpha3.Destination{
+				Host:   r.Destination.Host,
+				Port:   r.Destination.Port,
+				Subset: r.Destination.Subset + "-test",
```

Comment:
Let's make an issue to allow `-test` suffix be somehow configurable.

Reply:
'-test' should be some random code per session. maybe just session name?

Reply:
That would make sense - easier to collerate. :+1:

Reply:
https://github.com/aslakknutsen/istio-workspace/issues/58